pkg/distributor/forwarding/forwarding.go

```diff
@@ -96,185 +74,260 @@ func NewForwarder(reg prometheus.Registerer, cfg Config) Forwarder {
-		// to anywhere but we should still send it to the Ingesters.
-		return true
-	}
+// forwardingEndpointAssigner returns a func which takes a timeseries and assigns it endpoints to which it should be forwarded.
+func (f *forwarder) splitByTargets(tsSlice []mimirpb.PreallocTimeseries, rules validation.ForwardingRules) (TimeseriesCounts, []mimirpb.PreallocTimeseries, tsByTargets) {
+	// notIngestedCounts keeps track of the number of samples and exemplars that we don't send to the ingesters,
+	// we need to count these in order to later update some of the distributor's metrics correctly.
+	var notIngestedCounts TimeseriesCounts
+
+	// tsSliceWriteIdx is the index in the toIngest slice where we are writing TimeSeries to.
+	tsSliceWriteIdx := 0
+
+	tsByTargets := f.pools.getTsByTargets()
+	for tsSliceReadIdx, ts := range tsSlice {
+		ingest := false
+		var forwardingTarget string
+
+		metric, err := extract.UnsafeMetricNameFromLabelAdapters(ts.Labels)
+		if err != nil {
+			// Can't check whether a timeseries should be forwarded if it has no metric name.
+			// Send it to the Ingesters and don't forward it.
+			ingest = true
+		} else {
+			rule, ok := rules[metric]
+			if !ok {
+				// There is no forwarding rule for this metric, send it to the Ingesters and don't forward it.
+				ingest = true
+			} else {
+				forwardingTarget = rule.Endpoint
+				ingest = rule.Ingest
+			}
+		}
 
-	rule, ok := r.rules[metric]
-	if !ok {
-		// There is no forwarding rule for this metric, send it to the Ingesters.
-		return true
-	}
-	r.samples.Add(float64(len(sample.Samples)))
+		if forwardingTarget != "" {
+			tsByTarget, ok := tsByTargets[forwardingTarget]
+			if !ok {
+				tsByTarget.ts = f.pools.getTsSlice()
+			}
 
-	ts, ok := r.tsByEndpoint[rule.Endpoint]
-	if !ok {
-		ts = r.pools.timeseries.Get().(*[]mimirpb.PreallocTimeseries)
-		r.requests.Inc()
-	}
+			tsByTarget.ts = f.growTimeseriesSlice(tsByTarget.ts)
+			tsWriteIdx := len(tsByTarget.ts) - 1
+			mimirpb.DeepCopyTimeseries(tsByTarget.ts[tsWriteIdx].TimeSeries, ts.TimeSeries)
+			tsByTarget.counts.count(tsByTarget.ts[tsWriteIdx])
 
-	*ts = append(*ts, sample)
-	r.tsByEndpoint[rule.Endpoint] = ts
+			tsByTargets[forwardingTarget] = tsByTarget
+		}
 
-	return rule.Ingest
-}
+		if ingest {
+			if tsSliceWriteIdx != tsSliceReadIdx {
+				// Swap the timeseries at the reading and writing indices,
+				// later we'll return all timeseries beyond the write index to the pool.
+				tsSlice[tsSliceWriteIdx], tsSlice[tsSliceReadIdx] = tsSlice[tsSliceReadIdx], tsSlice[tsSliceWriteIdx]
+			}
+			tsSliceWriteIdx++
+		} else {
+			notIngestedCounts.count(ts)
+		}
+	}
+
+	// Truncate the toIngest slice to the index up to which we wrote TimeSeries data into it,
+	// all the TimeSeries objects beyond the write index must be returned to the pool.
+	for _, ts := range tsSlice[tsSliceWriteIdx:] {
+		f.pools.putTs(ts.TimeSeries)
+	}
+	tsSlice = tsSlice[:tsSliceWriteIdx]
```

Comment:
Even though all this code seems ok, IMO its complexity doesn't pay it's performance. I had to stop and think deeply about this `for` loop modifying the loop inside.

I'd consider just building a new slice with series to ingest, removing them from the original one, and then cleaning up the original one.
Getting a slice from the pool is almost zero price since we have lots of them already, so it won't be much worse in performance.

Reply:
I agree that unfortunately this is relatively complicated. The alternative which you're suggesting is a bit more complex than just getting a new slice from the pool and returning the old one because we also need to ensure that all of the `*mimirpb.TimeSeries` objects within the slice get returned to the pool and only once. This is simple if we deep copy them into the new slice, but I would prefer to avoid deep copying whenever possible because that's relatively expensive.

I implemented this here: 05c7e94e808220b52a0fb58db8c61ec5020db1e0

This is how the benchmarks look like before/after (with `-test.benchtime=3s`):

```
$ benchstat -delta-test none ./old ./new
name                                                      old time/op    new time/op    delta
RemoteWriteForwarding/forwarding_and_ingesting-4            1.63ms ± 0%    1.67ms ± 0%   +2.52%
RemoteWriteForwarding/forwarding_and_no_ingesting-4         1.65ms ± 0%    1.74ms ± 0%   +5.55%
RemoteWriteForwarding/no_forwarding_and_only_ingesting-4    1.38ms ± 0%    1.42ms ± 0%   +3.25%

name                                                      old alloc/op   new alloc/op   delta
RemoteWriteForwarding/forwarding_and_ingesting-4            2.74kB ± 0%    2.77kB ± 0%   +1.09%
RemoteWriteForwarding/forwarding_and_no_ingesting-4         3.02kB ± 0%    3.85kB ± 0%  +27.46%
RemoteWriteForwarding/no_forwarding_and_only_ingesting-4      183B ± 0%      209B ± 0%  +14.21%

name                                                      old allocs/op  new allocs/op  delta
RemoteWriteForwarding/forwarding_and_ingesting-4              29.0 ± 0%      30.0 ± 0%   +3.45%
RemoteWriteForwarding/forwarding_and_no_ingesting-4           30.0 ± 0%      33.0 ± 0%  +10.00%
RemoteWriteForwarding/no_forwarding_and_only_ingesting-4      4.00 ± 0%      5.00 ± 0%  +25.00%
```

Obviously it's hard to define how many allocations are "too much" relative to complexity of the code :) But I'd be fine to leave it like this.